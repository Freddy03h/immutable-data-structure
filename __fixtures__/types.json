[
  {
    "id": "106f524e-7283-44b8-aa84-25e9a7fb1f7d",
    "title": "Manga"
  },
  {
    "id": "fcb5dd8c-b28f-467d-9858-df52dd24e555",
    "title": "Manhwa"
  },
  {
    "id": "a4430b4d-4021-4a73-b286-6b76fb1bf103",
    "title": "Manhua"
  },
  {
    "id": "3c009f18-811d-4b07-8dde-8249e422ec9e",
    "title": "Manfra"
  },
  {
    "id": "bb3d92a9-020a-4d3f-adec-77bebe9980db",
    "title": "Roman"
  },
  {
    "id": "588f1a19-7a1e-498c-88f6-901d8a47e79f",
    "title": "GuideBook"
  },
  {
    "id": "f41d81b0-fca6-4445-a903-da310e0899d2",
    "title": "ArtBook"
  },
  {
    "id": "9c31ba97-aef8-455f-b6fd-35973874b9ad",
    "title": "BD"
  },
  {
    "id": "8f5eb0fd-a8d5-48ca-b694-0057102dd43c",
    "title": "Anime Comics"
  },
  {
    "id": "09253f84-8e89-4536-bfb2-bd91d9e0c0ab",
    "title": "Comics"
  }
]