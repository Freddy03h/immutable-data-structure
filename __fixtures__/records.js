import Immutable from 'immutable'

export const initialState = Immutable.Map({
  data: Immutable.Map(),
  relations: Immutable.Map(),
})

export const SerieRecord = Immutable.Record({
  id: null,
  title: null,
  type_id: null,
})
export const serieForeignKeys = ['type_id']

export const EditionRecord = Immutable.Record({
  id: null,
  title: null,
  publisher_id: null,
  series_id: null,
})
export const editionForeignKeys = ['publisher_id', 'series_id']
