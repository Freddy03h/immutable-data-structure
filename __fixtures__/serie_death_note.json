{
  "id": "320071be-4196-402b-98d7-d34bec8a1aab",
  "title": "Death Note",
  "type_id": "106f524e-7283-44b8-aa84-25e9a7fb1f7d",
  "tasks": [
    {
      "id": "63698905-4149-434a-abfd-59e2343c1c07",
      "series_id": "320071be-4196-402b-98d7-d34bec8a1aab",
      "author_id": "17491b54-580f-4d5f-94e5-9b9e501871c4",
      "job_id": "bf39f3fe-ce75-48b0-b557-877da935633b",
      "author": {
        "id": "17491b54-580f-4d5f-94e5-9b9e501871c4",
        "name": "Ōba",
        "first_name": "Tsugumi"
      }
    },
    {
      "id": "adb8225f-c8ee-41cc-9d14-2dd0b663a6cf",
      "series_id": "320071be-4196-402b-98d7-d34bec8a1aab",
      "author_id": "f06df71d-7bb3-4a4c-aec0-638f53aa43a8",
      "job_id": "37fb2477-ec0d-4ffc-9d3b-4189c2d01629",
      "author": {
        "id": "f06df71d-7bb3-4a4c-aec0-638f53aa43a8",
        "name": "Obata",
        "first_name": "Takeshi"
      }
    }
  ],
  "editions": [
    {
      "id": "908612fd-1cc7-4780-aea4-63d3ac04c6bd",
      "title": "Black Edition",
      "series_id": "320071be-4196-402b-98d7-d34bec8a1aab",
      "publisher_id": "4c9547ff-2ef6-439a-80b8-ea705a385b76",
      "volumes": [
        {
          "id": "ec9d0282-78a1-491e-bd1e-a937b332c56f",
          "title": "Tome 6",
          "number": 6,
          "date": "2011-05-06",
          "img": "https://images-eu.ssl-images-amazon.com/images/I/41FBONjMi4L.jpg",
          "edition_id": "908612fd-1cc7-4780-aea4-63d3ac04c6bd",
          "isbn": "9782505010951"
        },
        {
          "id": "ef282203-2ef5-4972-9598-d365fe7ff09c",
          "title": "Tome 5",
          "number": 5,
          "date": "2011-03-04",
          "img": "https://images-eu.ssl-images-amazon.com/images/I/41%2BDT%2B3-rwL.jpg",
          "edition_id": "908612fd-1cc7-4780-aea4-63d3ac04c6bd",
          "isbn": "9782505010692"
        },
        {
          "id": "69fed8ee-a500-4bde-9ba2-458f639785d1",
          "title": "Tome 4",
          "number": 4,
          "date": "2011-01-07",
          "img": "https://images-eu.ssl-images-amazon.com/images/I/41lRggimfJL.jpg",
          "edition_id": "908612fd-1cc7-4780-aea4-63d3ac04c6bd",
          "isbn": "9782505010326"
        },
        {
          "id": "7ac3f560-bb02-4a9e-9e01-67cc2e2b9d32",
          "title": "Tome 3",
          "number": 3,
          "date": "2010-11-05",
          "img": "https://images-eu.ssl-images-amazon.com/images/I/41mLoqm6D0L.jpg",
          "edition_id": "908612fd-1cc7-4780-aea4-63d3ac04c6bd",
          "isbn": "9782505009986"
        },
        {
          "id": "3a03733c-1a5a-4673-9299-28e5cc82a8af",
          "title": "Tome 2",
          "number": 2,
          "date": "2010-09-03",
          "img": "https://images-eu.ssl-images-amazon.com/images/I/51KMP3phtGL.jpg",
          "edition_id": "908612fd-1cc7-4780-aea4-63d3ac04c6bd",
          "isbn": "9782505009061"
        },
        {
          "id": "dbab2483-2d3c-4832-9993-b6efbd765787",
          "title": "Tome 1",
          "number": 1,
          "date": "2010-07-02",
          "img": "https://images-eu.ssl-images-amazon.com/images/I/413rEPunmPL.jpg",
          "edition_id": "908612fd-1cc7-4780-aea4-63d3ac04c6bd",
          "isbn": "9782505009054"
        }
      ]
    },
    {
      "id": "5482eba9-2773-40d8-82f7-1870b1b2a3e9",
      "title": "Edition simple",
      "series_id": "320071be-4196-402b-98d7-d34bec8a1aab",
      "publisher_id": "4c9547ff-2ef6-439a-80b8-ea705a385b76",
      "volumes": [
        {
          "id": "081648ff-1b14-4af7-b8f0-9e97b0ba6804",
          "title": "Tome 13",
          "number": 13,
          "date": "2009-09-18",
          "img": "https://images-eu.ssl-images-amazon.com/images/I/51JrRDR1ANL.jpg",
          "edition_id": "5482eba9-2773-40d8-82f7-1870b1b2a3e9",
          "isbn": "9782505004523"
        },
        {
          "id": "11606764-7aee-4e33-9df1-e9dc5298eda8",
          "title": "Tome 12",
          "number": 12,
          "date": "2008-10-02",
          "img": "https://images-eu.ssl-images-amazon.com/images/I/61XFUTPtQdL.jpg",
          "edition_id": "5482eba9-2773-40d8-82f7-1870b1b2a3e9",
          "isbn": "9782505004295"
        },
        {
          "id": "a4477f49-cff1-4913-bd39-68fefbc10479",
          "title": "Tome 11",
          "number": 11,
          "date": "2008-07-04",
          "img": "https://images-eu.ssl-images-amazon.com/images/I/618b1GXgGFL.jpg",
          "edition_id": "5482eba9-2773-40d8-82f7-1870b1b2a3e9",
          "isbn": "9782505003694"
        },
        {
          "id": "c1995831-6447-43e9-869b-74f2cf93966d",
          "title": "Tome 10",
          "number": 10,
          "date": "2008-06-06",
          "img": "https://images-eu.ssl-images-amazon.com/images/I/61OEH0gsrIL.jpg",
          "edition_id": "5482eba9-2773-40d8-82f7-1870b1b2a3e9",
          "isbn": "9782505003038"
        },
        {
          "id": "73f6690b-716b-4659-97f5-d66ce10acd8a",
          "title": "Tome 9",
          "number": 9,
          "date": "2008-04-11",
          "img": "https://images-eu.ssl-images-amazon.com/images/I/61jqfMbZ8VL.jpg",
          "edition_id": "5482eba9-2773-40d8-82f7-1870b1b2a3e9",
          "isbn": "9782505002970"
        },
        {
          "id": "12b5dc0d-253a-41dd-bbee-e1f31e947eb1",
          "title": "Tome 8",
          "number": 8,
          "date": "2008-02-15",
          "img": "https://images-eu.ssl-images-amazon.com/images/I/51ZlUQFdaPL.jpg",
          "edition_id": "5482eba9-2773-40d8-82f7-1870b1b2a3e9",
          "isbn": "9782505002697"
        },
        {
          "id": "f0bc0f76-bddf-4af8-a589-0d877aae9235",
          "title": "Tome 7",
          "number": 7,
          "date": "2007-12-07",
          "img": "https://images-eu.ssl-images-amazon.com/images/I/61qgonxtZOL.jpg",
          "edition_id": "5482eba9-2773-40d8-82f7-1870b1b2a3e9",
          "isbn": "9782505001997"
        },
        {
          "id": "a2280876-7082-48b5-bab3-34c4e5eab279",
          "title": "Tome 6",
          "number": 6,
          "date": "2007-10-05",
          "img": "https://images-eu.ssl-images-amazon.com/images/I/51ccrw7ky1L.jpg",
          "edition_id": "5482eba9-2773-40d8-82f7-1870b1b2a3e9",
          "isbn": "9782505001812"
        },
        {
          "id": "b95ab092-9baf-4feb-b8f1-e79b449b9c54",
          "title": "Tome 5",
          "number": 5,
          "date": "2007-08-24",
          "img": "https://images-eu.ssl-images-amazon.com/images/I/61bXKQlOllL.jpg",
          "edition_id": "5482eba9-2773-40d8-82f7-1870b1b2a3e9",
          "isbn": "9782505001614"
        },
        {
          "id": "e7dbccb5-4710-446b-8c08-d35fb8a1eaeb",
          "title": "Tome 4",
          "number": 4,
          "date": "2007-05-31",
          "img": "https://images-eu.ssl-images-amazon.com/images/I/611d8aPhqhL.jpg",
          "edition_id": "5482eba9-2773-40d8-82f7-1870b1b2a3e9",
          "isbn": "9782505001065"
        },
        {
          "id": "fa7a1162-d9e3-462f-a367-8371a55f7c94",
          "title": "Tome 3",
          "number": 3,
          "date": "2007-04-05",
          "img": "https://images-eu.ssl-images-amazon.com/images/I/61fiEjVbUbL.jpg",
          "edition_id": "5482eba9-2773-40d8-82f7-1870b1b2a3e9",
          "isbn": "9782505000792"
        },
        {
          "id": "11d1491e-22d4-49f1-b29f-78007493d4e6",
          "title": "Tome 2",
          "number": 2,
          "date": "2007-02-01",
          "img": "https://images-eu.ssl-images-amazon.com/images/I/61DduSmdKGL.jpg",
          "edition_id": "5482eba9-2773-40d8-82f7-1870b1b2a3e9",
          "isbn": "9782505000426"
        },
        {
          "id": "ff15ae7e-fcfe-4061-a27b-3b481e5f313c",
          "title": "Tome 1",
          "number": 1,
          "date": "2007-01-18",
          "img": "https://images-eu.ssl-images-amazon.com/images/I/51eBp6XUd5L.jpg",
          "edition_id": "5482eba9-2773-40d8-82f7-1870b1b2a3e9",
          "isbn": "9782505000327"
        }
      ]
    }
  ],
  "kinds": [
    {
      "id": "9d6a131c-4b4e-4cf6-a9aa-de78087d672d",
      "title": "Fantastique"
    },
    {
      "id": "c67e1d91-20c3-43b6-98da-aea886176da0",
      "title": "Gothique"
    },
    {
      "id": "eb5eb1b3-9f89-43ab-b0fa-0ef61f10d633",
      "title": "Enquête"
    },
    {
      "id": "acd6bec9-507d-46ab-a1de-f18f8b8ca83c",
      "title": "Drame"
    },
    {
      "id": "20840e27-413b-47c1-83f3-042be2c5ff0d",
      "title": "Dieu et Déesse"
    },
    {
      "id": "5faba927-085f-4fbd-a468-4456352ba872",
      "title": "Psychologique"
    },
    {
      "id": "d806ffd0-c43e-4e78-a557-8d7b41cc44e2",
      "title": "Assassin"
    },
    {
      "id": "eb84f0d1-680d-4127-a416-3bb78c365d9d",
      "title": "Agent secret"
    },
    {
      "id": "e0919979-ca6f-4025-ba1d-9897d5f075d4",
      "title": "Thriller"
    },
    {
      "id": "d5a3385a-cfcd-45f6-8f82-0c80956b084d",
      "title": "Shōnen"
    },
    {
      "id": "fa3722c2-f18c-480e-bd4d-e915de81e69d",
      "title": "Policier"
    }
  ]
}