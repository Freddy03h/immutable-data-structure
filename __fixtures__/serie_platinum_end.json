{
  "id": "fc34ad95-27cf-4cba-aafc-e9c37de532b4",
  "title": "Platinum End",
  "type_id": "106f524e-7283-44b8-aa84-25e9a7fb1f7d",
  "possessions_count": 1024,
  "tasks": [
    {
      "id": "4b1d117d-063b-4a08-b6d7-2f3d252b2a81",
      "series_id": "fc34ad95-27cf-4cba-aafc-e9c37de532b4",
      "author_id": "f06df71d-7bb3-4a4c-aec0-638f53aa43a8",
      "job_id": "37fb2477-ec0d-4ffc-9d3b-4189c2d01629",
      "author": {
        "id": "f06df71d-7bb3-4a4c-aec0-638f53aa43a8",
        "name": "Obata",
        "first_name": "Takeshi"
      }
    },
    {
      "id": "5195c5d1-9d3e-4a6c-9b29-077fc217d224",
      "series_id": "fc34ad95-27cf-4cba-aafc-e9c37de532b4",
      "author_id": "17491b54-580f-4d5f-94e5-9b9e501871c4",
      "job_id": "bf39f3fe-ce75-48b0-b557-877da935633b",
      "author": {
        "id": "17491b54-580f-4d5f-94e5-9b9e501871c4",
        "name": "Ōba",
        "first_name": "Tsugumi"
      }
    }
  ],
  "editions": [
    {
      "id": "73173187-5965-451c-adbe-84e3e719abc0",
      "title": "Edition simple",
      "series_id": "fc34ad95-27cf-4cba-aafc-e9c37de532b4",
      "publisher_id": "122e0ec3-f072-4c5d-b40d-3eba4a82fe1e",
      "volumes": [
        {
          "id": "27e50b0a-785e-4298-af2f-ae492c57c8a6",
          "title": "Tome 6",
          "number": 6,
          "date": "2017-09-20",
          "img": "",
          "edition_id": "73173187-5965-451c-adbe-84e3e719abc0",
          "isbn": "9782820329011"
        },
        {
          "id": "f6f21458-e707-40ba-b9ec-f7f28d7c6343",
          "title": "Tome 5",
          "number": 5,
          "date": "2017-05-31",
          "img": "https://images-eu.ssl-images-amazon.com/images/I/61CV%2BQwDT7L.jpg",
          "edition_id": "73173187-5965-451c-adbe-84e3e719abc0",
          "isbn": "9782820328533"
        },
        {
          "id": "06fe038b-8c0d-4767-9167-ae3ca9bc3067",
          "title": "Tome 4",
          "number": 4,
          "date": "2017-03-08",
          "img": "https://images-eu.ssl-images-amazon.com/images/I/51tKK5n%2B-3L.jpg",
          "edition_id": "73173187-5965-451c-adbe-84e3e719abc0",
          "isbn": "9782820328175"
        },
        {
          "id": "7c0ad1a9-4a80-4710-b11b-eed249456924",
          "title": "Tome 3",
          "number": 3,
          "date": "2016-11-23",
          "img": "https://images-eu.ssl-images-amazon.com/images/I/51jAjenb0EL.jpg",
          "edition_id": "73173187-5965-451c-adbe-84e3e719abc0",
          "isbn": "9782820325402"
        },
        {
          "id": "975e5e39-8a47-4e0c-9a60-4a6256163257",
          "title": "Tome 2",
          "number": 2,
          "date": "2016-09-07",
          "img": "https://images-eu.ssl-images-amazon.com/images/I/51sgTWp8W-L.jpg",
          "edition_id": "73173187-5965-451c-adbe-84e3e719abc0",
          "isbn": "9782820324948"
        },
        {
          "id": "04754705-044e-4292-8ed2-eee50ee91d2d",
          "title": "Tome 1",
          "number": 1,
          "date": "2016-05-25",
          "img": "https://images-eu.ssl-images-amazon.com/images/I/51TgELYk1WL.jpg",
          "edition_id": "73173187-5965-451c-adbe-84e3e719abc0",
          "isbn": "9782820324603"
        }
      ]
    }
  ],
  "kinds": [
    {
      "id": "d5a3385a-cfcd-45f6-8f82-0c80956b084d",
      "title": "Shōnen"
    },
    {
      "id": "9d6a131c-4b4e-4cf6-a9aa-de78087d672d",
      "title": "Fantastique"
    },
    {
      "id": "acd6bec9-507d-46ab-a1de-f18f8b8ca83c",
      "title": "Drame"
    }
  ]
}