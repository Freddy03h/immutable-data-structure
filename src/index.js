import Immutable from 'immutable'

const mapListToTuple = (list, key, record) => list.map((item) => [item.get(key), record ? new record(item) : item])

export const listToMapWithKey = (list, key) => Immutable.Map(mapListToTuple(list, key))
export const listToOrderedMapWithKey = (list, key) => Immutable.OrderedMap(mapListToTuple(list, key))

export const arrayToMapWithKey = (array, key) => listToMapWithKey(Immutable.fromJS(array), key)
export const arrayToOrderedMapWithKey = (array, key) => listToOrderedMapWithKey(Immutable.fromJS(array), key)

export const listToMapWithKeyAndRecord = (list, key, record) => Immutable.Map(mapListToTuple(list, key, record))
export const listToOrderedMapWithKeyAndRecord = (list, key, record) => Immutable.OrderedMap(mapListToTuple(list, key, record))

// predicate function for Immutable
export function keyIn(keys) {
  const keySet = Immutable.Set(keys)
  return (v, k) => keySet.has(k)
}

export const listGroupToMapByForeignKey = (list, foreign_id, key = 'id') => {
  return list.reduce((accumulator, item) => {
    if (!item.get(foreign_id)) {
      return accumulator
    }
    return accumulator.update(
      item.get(foreign_id),
      Immutable.Set(),
      (item_ids) => item_ids.add(item.get(key)),
    )
  }, Immutable.Map())
}

export const listGroupToMapByForeignKeys = (list, foreign_ids, key = 'id') => {
  return list.reduce((accumulator, item) => {
    return item.get(foreign_ids).reduce((acc, foreign_id) => {
      return acc.update(
        foreign_id,
        Immutable.Set(),
        (item_ids) => item_ids.add(item.get(key)),
      )
    }, accumulator)
  }, Immutable.Map())
}

export const getDataByForeignId = (store, moduleName, foreignIdKey, foreignId, key = 'id') => {
  const relatedItemIds = store.getIn([moduleName, 'relations', foreignIdKey, foreignId], Immutable.Map())
  const listRelatedItemIds = relatedItemIds.map((itemId) => store.getIn([moduleName, 'data', itemId])).toList()

  return listToMapWithKey(listRelatedItemIds, key)
}

export const getDataByForeignIds = (store, moduleName, foreignIdKey, foreignIds, key = 'id') => {
  const relatedItemIds = foreignIds
    ? foreignIds.reduce((accumulator, otherItemId) => {
        const itemIds = store.getIn([moduleName, 'relations', foreignIdKey, otherItemId])
        return itemIds ? accumulator.union(itemIds) : accumulator
      }, Immutable.Set())
    : Immutable.Set()

  const listRelatedItemIds = relatedItemIds
    ? relatedItemIds.map((itemId) => store.getIn([moduleName, 'data', itemId])).toList()
    : Immutable.List()

  return listToMapWithKey(listRelatedItemIds, key)
}

export const getDataByForeignIdThroughOtherForeignId = (store, moduleName, foreignIdKey, otherModuleName, otherForeignIdKey, otherForeignId, key = 'id') => {
  const otherForeignIds = store.getIn([otherModuleName, 'relations', otherForeignIdKey, otherForeignId])

  return getDataByForeignIds(store, moduleName, foreignIdKey, otherForeignIds, key)
}

//////

export const updateRecord = (store, Record, newData, foreignKeys = [], primaryKey = 'id') => {
  const id = newData.get(primaryKey)
  const oldData = store.getIn(['data', id])

  return store.updateIn(['data', id], new Record(), (record) => record.merge(newData))
    .withMutations((collection) => {
      foreignKeys.forEach(
        (foreignKey) => {
          if(oldData) {
            collection.updateIn(['relations', foreignKey, oldData.get(foreignKey)], (item_ids) => item_ids && item_ids.remove(id))
          }
          if(newData.get(foreignKey)) {
            collection.updateIn(['relations', foreignKey, newData.get(foreignKey)], new Immutable.Set(), (item_ids) => item_ids.add(id))
          }
        }
      )
    })
}

export const deleteRecord = (store, id, foreignKeys = []) => {
  const data = store.getIn(['data', id])

  if(!data) {
    return store
  }

  return store.deleteIn(['data', id])
    .withMutations((collection) => {
      foreignKeys.forEach(
        (foreignKey) => {
          collection.updateIn(['relations', foreignKey, data.get(foreignKey)], (item_ids) => item_ids && item_ids.remove(id))
        }
      )
    })
}

export const mergeRecords = (store, Record, listData, foreignKeys = [], primaryKey = 'id', completeKeysPath = null) => {
  let keysToRemove = null
  if(completeKeysPath) {
    const completeKeys = store.getIn(['relations', ...completeKeysPath])
    if(completeKeys) {
      const actuelKeys = listData ? listData.map((mapItem) => mapItem.get(primaryKey)).toSet() : Immutable.Set()

      keysToRemove = completeKeys.subtract(actuelKeys)
    }
  }

  return store
    .withMutations((collection) => {
      if(listData) {
        listData.forEach(
          (mapItem) => {
            updateRecord(collection, Record, mapItem, foreignKeys, primaryKey)
          }
        )
      }

      if(keysToRemove) {
        keysToRemove.forEach(
          (id) => {
            deleteRecord(collection, id, foreignKeys)
          }
        )
      }
    })
}

export const mergeCompleteListsRecords = (store, Record, listData, keyData, relationKey, keyDataPrimary, foreignKeys = [], primaryKey = 'id') => {
  return store
    .withMutations((collection) => {
      listData.forEach(
        (mapItem) => {
          mergeRecords(collection, Record, mapItem.get(keyData), foreignKeys, primaryKey, [relationKey, mapItem.get(keyDataPrimary)])
        }
      )
    })
}

