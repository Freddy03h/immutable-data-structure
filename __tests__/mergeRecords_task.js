import Immutable from 'immutable'
import { initialState, createMergeRecords } from '../src/index'

import authorObaJSON from '../__fixtures__/author_oba.json'
import seriePlatinumEndJSON from '../__fixtures__/serie_platinum_end.json'

import { TaskRecord, taskForeignKeys } from '../__fixtures__/records'

const mergeTasksRecords = createMergeRecords(TaskRecord, taskForeignKeys)

const author = Immutable.fromJS(authorObaJSON)
const serie = Immutable.fromJS(seriePlatinumEndJSON)

describe('mergeRecords tasks', () => {

  test('one author with empty store', () => {
    expect(
      mergeTasksRecords(initialState, author.get('tasks'), ['author_id', author.get('id')])
    ).toMatchSnapshot()
  })

  test('one author then serie with empty store', () => {
    const store = mergeTasksRecords(initialState, author.get('tasks'), ['author_id', author.get('id')])
    expect(
      mergeTasksRecords(store, serie.get('tasks'), ['series_id', serie.get('id')])
    ).toMatchSnapshot()
  })

  test('one author change task order', () => {
    const authorEdited = author.updateIn(['tasks'], (list) => list.sortBy((item) => item.get('id')))
    const store = mergeTasksRecords(initialState, authorEdited.get('tasks'), ['author_id', authorEdited.get('id')])

    expect(
      mergeTasksRecords(store, author.get('tasks'), ['author_id', author.get('id')])
    ).toMatchSnapshot()
  })

})
