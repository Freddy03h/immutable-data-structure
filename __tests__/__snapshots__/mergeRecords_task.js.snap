// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mergeRecords tasks one author change task order 1`] = `
Immutable.Map {
  "data": Immutable.Map {
    "0b6a3f23-ed97-46a6-b8bb-25ced41d9442": Immutable.Record {
      "id": "0b6a3f23-ed97-46a6-b8bb-25ced41d9442",
      "author_id": "17491b54-580f-4d5f-94e5-9b9e501871c4",
      "job_id": "bf39f3fe-ce75-48b0-b557-877da935633b",
      "series_id": "277484f9-d839-4b44-9f43-0743e6db540f",
    },
    "5195c5d1-9d3e-4a6c-9b29-077fc217d224": Immutable.Record {
      "id": "5195c5d1-9d3e-4a6c-9b29-077fc217d224",
      "author_id": "17491b54-580f-4d5f-94e5-9b9e501871c4",
      "job_id": "bf39f3fe-ce75-48b0-b557-877da935633b",
      "series_id": "fc34ad95-27cf-4cba-aafc-e9c37de532b4",
    },
    "63698905-4149-434a-abfd-59e2343c1c07": Immutable.Record {
      "id": "63698905-4149-434a-abfd-59e2343c1c07",
      "author_id": "17491b54-580f-4d5f-94e5-9b9e501871c4",
      "job_id": "bf39f3fe-ce75-48b0-b557-877da935633b",
      "series_id": "320071be-4196-402b-98d7-d34bec8a1aab",
    },
  },
  "relations": Immutable.Map {
    "author_id": Immutable.Map {
      "17491b54-580f-4d5f-94e5-9b9e501871c4": Immutable.OrderedSet [
        "5195c5d1-9d3e-4a6c-9b29-077fc217d224",
        "0b6a3f23-ed97-46a6-b8bb-25ced41d9442",
        "63698905-4149-434a-abfd-59e2343c1c07",
      ],
    },
    "job_id": Immutable.Map {
      "bf39f3fe-ce75-48b0-b557-877da935633b": Immutable.OrderedSet [
        "0b6a3f23-ed97-46a6-b8bb-25ced41d9442",
        "5195c5d1-9d3e-4a6c-9b29-077fc217d224",
        "63698905-4149-434a-abfd-59e2343c1c07",
      ],
    },
    "series_id": Immutable.Map {
      "277484f9-d839-4b44-9f43-0743e6db540f": Immutable.OrderedSet [
        "0b6a3f23-ed97-46a6-b8bb-25ced41d9442",
      ],
      "fc34ad95-27cf-4cba-aafc-e9c37de532b4": Immutable.OrderedSet [
        "5195c5d1-9d3e-4a6c-9b29-077fc217d224",
      ],
      "320071be-4196-402b-98d7-d34bec8a1aab": Immutable.OrderedSet [
        "63698905-4149-434a-abfd-59e2343c1c07",
      ],
    },
  },
}
`;

exports[`mergeRecords tasks one author then serie with empty store 1`] = `
Immutable.Map {
  "data": Immutable.Map {
    "5195c5d1-9d3e-4a6c-9b29-077fc217d224": Immutable.Record {
      "id": "5195c5d1-9d3e-4a6c-9b29-077fc217d224",
      "author_id": "17491b54-580f-4d5f-94e5-9b9e501871c4",
      "job_id": "bf39f3fe-ce75-48b0-b557-877da935633b",
      "series_id": "fc34ad95-27cf-4cba-aafc-e9c37de532b4",
    },
    "0b6a3f23-ed97-46a6-b8bb-25ced41d9442": Immutable.Record {
      "id": "0b6a3f23-ed97-46a6-b8bb-25ced41d9442",
      "author_id": "17491b54-580f-4d5f-94e5-9b9e501871c4",
      "job_id": "bf39f3fe-ce75-48b0-b557-877da935633b",
      "series_id": "277484f9-d839-4b44-9f43-0743e6db540f",
    },
    "63698905-4149-434a-abfd-59e2343c1c07": Immutable.Record {
      "id": "63698905-4149-434a-abfd-59e2343c1c07",
      "author_id": "17491b54-580f-4d5f-94e5-9b9e501871c4",
      "job_id": "bf39f3fe-ce75-48b0-b557-877da935633b",
      "series_id": "320071be-4196-402b-98d7-d34bec8a1aab",
    },
    "4b1d117d-063b-4a08-b6d7-2f3d252b2a81": Immutable.Record {
      "id": "4b1d117d-063b-4a08-b6d7-2f3d252b2a81",
      "author_id": "f06df71d-7bb3-4a4c-aec0-638f53aa43a8",
      "job_id": "37fb2477-ec0d-4ffc-9d3b-4189c2d01629",
      "series_id": "fc34ad95-27cf-4cba-aafc-e9c37de532b4",
    },
  },
  "relations": Immutable.Map {
    "author_id": Immutable.Map {
      "17491b54-580f-4d5f-94e5-9b9e501871c4": Immutable.OrderedSet [
        "5195c5d1-9d3e-4a6c-9b29-077fc217d224",
        "0b6a3f23-ed97-46a6-b8bb-25ced41d9442",
        "63698905-4149-434a-abfd-59e2343c1c07",
      ],
      "f06df71d-7bb3-4a4c-aec0-638f53aa43a8": Immutable.OrderedSet [
        "4b1d117d-063b-4a08-b6d7-2f3d252b2a81",
      ],
    },
    "job_id": Immutable.Map {
      "bf39f3fe-ce75-48b0-b557-877da935633b": Immutable.OrderedSet [
        "5195c5d1-9d3e-4a6c-9b29-077fc217d224",
        "0b6a3f23-ed97-46a6-b8bb-25ced41d9442",
        "63698905-4149-434a-abfd-59e2343c1c07",
      ],
      "37fb2477-ec0d-4ffc-9d3b-4189c2d01629": Immutable.OrderedSet [
        "4b1d117d-063b-4a08-b6d7-2f3d252b2a81",
      ],
    },
    "series_id": Immutable.Map {
      "fc34ad95-27cf-4cba-aafc-e9c37de532b4": Immutable.OrderedSet [
        "4b1d117d-063b-4a08-b6d7-2f3d252b2a81",
        "5195c5d1-9d3e-4a6c-9b29-077fc217d224",
      ],
      "277484f9-d839-4b44-9f43-0743e6db540f": Immutable.OrderedSet [
        "0b6a3f23-ed97-46a6-b8bb-25ced41d9442",
      ],
      "320071be-4196-402b-98d7-d34bec8a1aab": Immutable.OrderedSet [
        "63698905-4149-434a-abfd-59e2343c1c07",
      ],
    },
  },
}
`;

exports[`mergeRecords tasks one author with empty store 1`] = `
Immutable.Map {
  "data": Immutable.Map {
    "5195c5d1-9d3e-4a6c-9b29-077fc217d224": Immutable.Record {
      "id": "5195c5d1-9d3e-4a6c-9b29-077fc217d224",
      "author_id": "17491b54-580f-4d5f-94e5-9b9e501871c4",
      "job_id": "bf39f3fe-ce75-48b0-b557-877da935633b",
      "series_id": "fc34ad95-27cf-4cba-aafc-e9c37de532b4",
    },
    "0b6a3f23-ed97-46a6-b8bb-25ced41d9442": Immutable.Record {
      "id": "0b6a3f23-ed97-46a6-b8bb-25ced41d9442",
      "author_id": "17491b54-580f-4d5f-94e5-9b9e501871c4",
      "job_id": "bf39f3fe-ce75-48b0-b557-877da935633b",
      "series_id": "277484f9-d839-4b44-9f43-0743e6db540f",
    },
    "63698905-4149-434a-abfd-59e2343c1c07": Immutable.Record {
      "id": "63698905-4149-434a-abfd-59e2343c1c07",
      "author_id": "17491b54-580f-4d5f-94e5-9b9e501871c4",
      "job_id": "bf39f3fe-ce75-48b0-b557-877da935633b",
      "series_id": "320071be-4196-402b-98d7-d34bec8a1aab",
    },
  },
  "relations": Immutable.Map {
    "author_id": Immutable.Map {
      "17491b54-580f-4d5f-94e5-9b9e501871c4": Immutable.OrderedSet [
        "5195c5d1-9d3e-4a6c-9b29-077fc217d224",
        "0b6a3f23-ed97-46a6-b8bb-25ced41d9442",
        "63698905-4149-434a-abfd-59e2343c1c07",
      ],
    },
    "job_id": Immutable.Map {
      "bf39f3fe-ce75-48b0-b557-877da935633b": Immutable.OrderedSet [
        "5195c5d1-9d3e-4a6c-9b29-077fc217d224",
        "0b6a3f23-ed97-46a6-b8bb-25ced41d9442",
        "63698905-4149-434a-abfd-59e2343c1c07",
      ],
    },
    "series_id": Immutable.Map {
      "fc34ad95-27cf-4cba-aafc-e9c37de532b4": Immutable.OrderedSet [
        "5195c5d1-9d3e-4a6c-9b29-077fc217d224",
      ],
      "277484f9-d839-4b44-9f43-0743e6db540f": Immutable.OrderedSet [
        "0b6a3f23-ed97-46a6-b8bb-25ced41d9442",
      ],
      "320071be-4196-402b-98d7-d34bec8a1aab": Immutable.OrderedSet [
        "63698905-4149-434a-abfd-59e2343c1c07",
      ],
    },
  },
}
`;
